File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        if(head ==null) return false;
        Set<Integer> visited = new HashSet<>();
        ListNode cur = head;
        visited.add(cur.hashCode());
        while(cur.next != null){
            if(visited.contains(cur.next.hashCode())) return true; 
            visited.add(cur.next.hashCode());
            cur = cur.next;
        }
        return false;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func mergeKLists(lists []*ListNode) *ListNode {
    var d ListNode
	cur := &d
	m := make(map[int]*ListNode)
	for i,l := range lists {
		if l == nil {
			continue
		}
		m[i] = l	
	}
	for len(m)>0 {
		minLoop := 1001
		contributor := -1
		for i, v := range m {
			if v.Val <= minLoop {
				minLoop = v.Val
				contributor = i
			}
			
		}
		cur.Next = m[contributor]
		cur = cur.Next
		m[contributor] = m[contributor].Next
		if m[contributor] == nil {
				delete(m,contributor)
		}
	}
	return d.Next
}
